### ISSUE TITLE
Class  and loss problem training with custom data

### ISSUE THREAD

**Issue by Antoine-ls** [zylo117/Yet-Another-EfficientDet-Pytorch#54]
@zylo117 First, thanks a lot to your great work. I want to train on custom data with only one class, while it's not coco formatted. I have changed yml file and its like:
![image](https://user-images.githubusercontent.com/40172933/79116112-308a9900-7dba-11ea-8a7a-4ce1089b2673.png)

If I set its class id in get_item function in my dataset as 1, it will trigger device-side error, which is normal case when class number and actual number are mismatched. If I change it to 0, the code can run, is there something like background id in your implementation?  I want to know if I am right to set it as 0..

Furthermore, I can actually run with class id 0 but I find that classification loss is too big, more than 1000 at beginning and decrease to 10 after 4 epochs, I don't know whether it's normal because classicification loss dominates the loss.

![image](https://user-images.githubusercontent.com/40172933/79115875-988caf80-7db9-11ea-9c1b-20a3248ab9b4.png)

Thanks in advance if you can answer this question.

**Comment by mad-fogs** [comment]
i also met this issue.
the training loss turned to be abnormal after several epoches.

**Comment by zylo117** [comment]
1, if you have to change the code to manually assign class index, your dataset convertion is probably unsuccessful.

2, reduce lr by 100 times and try again

**Comment by Antoine-ls** [comment]
In fact, I implement my own dataset function but with same return format, by class index I mean in the annots, it's [x1,y1,x2,y2,class] formatted, If I have only one class to detect, I should put 0 or 1 for this class.
Because in coco, I think 0 means background, so it's 1 for coco maybe. but if I put 1 in my case, and my obj_list has only one class which is 'abnormal', there will be device-side error as I mentioned before.

**Comment by zylo117** [comment]
> In fact, I implement my own dataset function but with same return format, by class index I mean in the annots, it's [x1,y1,x2,y2,class] formatted, If I have only one class to detect, I should put 0 or 1 for this class.
> Because in coco, I think 0 means background, so it's 1 for coco maybe. but if I put 1 in my case, and my obj_list has only one class which is 'abnormal', there will be device-side error as I mentioned before.

So the problem came from your dataset loader.

There is no 0 in coco and there is no 'backgrond' in coco. Also there is no background class in effdet. If you implement a custom dataset, it should contain only one class, and the index should be 0.

**Comment by betterhalfwzm** [comment]
@zylo117  I set my dataset one class，category_id = 1，is it ok?

**Comment by zylo117** [comment]
yes

**Comment by tfsong** [comment]
Thank you very much for your work！
 I tried to write a txt data reading interface about darknet-yolo. When training a class, the classification loss is very large. Where does this problem generally occur?

**Comment by zylo117** [comment]
> Thank you very much for your work！
> I tried to write a txt data reading interface about darknet-yolo. When training a class, the classification loss is very large. Where does this problem generally occur?

It takes time, keep training

**Comment by praneet195** [comment]
So can we add negative samples while training as background or can we only train on positive samples i.e images with objects?

**Comment by zylo117** [comment]
> So can we add negative samples while training as background or can we only train on positive samples i.e images with objects?

sure, just create a dataset with an extra class named background, you can label the background by yourself under that class
